parser-typechecker/src/Unison/Codebase/FileCodebase.hs

```diff
@@ -162,27 +164,30 @@ codebase1
    . MonadUnliftIO m
   => Var v
   => BuiltinAnnotation a
-  => S.Format v -> S.Format a -> CodebasePath -> Codebase m v a
+  => Branch.Cache m -> S.Format v -> S.Format a -> CodebasePath -> m (Codebase m v a)
 codebase1 = codebase1' Sync.syncToDirectory
 
 codebase1'
   :: forall m v a
    . MonadUnliftIO m
   => Var v
   => BuiltinAnnotation a
-  => Common.SyncToDir m v a -> S.Format v -> S.Format a -> CodebasePath -> Codebase m v a
-codebase1' syncToDirectory fmtV@(S.Format getV putV) fmtA@(S.Format getA putA) path =
+  => Common.SyncToDir m v a -> Branch.Cache m -> S.Format v -> S.Format a -> CodebasePath -> m (Codebase m v a)
+codebase1' syncToDirectory branchCache fmtV@(S.Format getV putV) fmtA@(S.Format getA putA) path = do
+  termCache <- Cache.semispaceCache 8192
+  typeOfTermCache <- Cache.semispaceCache 8192
+  declCache <- Cache.semispaceCache 1024
   let c =
         Codebase
-          (getTerm getV getA path)
-          (getTypeOfTerm getV getA path)
-          (getDecl getV getA path)
+          (Cache.applyDefined termCache $ getTerm getV getA path)
+          (Cache.applyDefined typeOfTermCache $ getTypeOfTerm getV getA path)
+          (Cache.applyDefined declCache $ getDecl getV getA path)
```

Comment:
`Cache.applyDefined` only caches results that are non-empty, which is the right move here - if you lookup a term in the cache and it isn't found now, you don't want to cache that it's missing, since it might be found later (say, after a `pull`).